File: src/main/java/org/seancheer/twopointers/_15_3Sum.java
package org.seancheer.twopointers;

import org.w3c.dom.html.HTMLPreElement;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
 * <p>
 * Notice that the solution set must not contain duplicate triplets.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: nums = [-1,0,1,2,-1,-4]
 * Output: [[-1,-1,2],[-1,0,1]]
 * Example 2:
 * <p>
 * Input: nums = []
 * Output: []
 * Example 3:
 * <p>
 * Input: nums = [0]
 * Output: []
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 0 <= nums.length <= 3000
 * -105 <= nums[i] <= 105
 *
 * @author: seancheer
 * @date: 2021/8/13
 **/
public class _15_3Sum {

    public static void main(String[] args) {
        var obj = new _15_3Sum();
        int[] nums = new int[]{-1, 0, 1, 2, -1, -4};
        var res = obj.threeSum(nums);
        var res2 = obj.threeSum2(nums);
        System.out.println("res:" + res + "   res2:" + res2);//[[-1,-1,2],[-1,0,1]]


        nums = new int[]{0};
        res = obj.threeSum(nums);
        res2 = obj.threeSum2(nums);
        System.out.println("res:" + res + "   res2:" + res2);

        nums = new int[]{-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4};
        res = obj.threeSum(nums);
        res2 = obj.threeSum2(nums);
        System.out.println("res:" + res + "   res2:" + res2);//[[-4,0,4],[-4,1,3],[-3,-1,4],[-3,0,3],[-3,1,2],[-2,-1,3],[-2,0,2],[-1,-1,2],[-1,0,1]]

        nums = new int[]{0,0,0};
        res = obj.threeSum(nums);
        res2 = obj.threeSum2(nums);
        System.out.println("res:" + res + "   res2:" + res2);//[[0,0,0]]

    }

    /**
     * 解题思路：使用双指针来确定前两个数，最后一个数使用二分法来查找加快速度
     *
     * @param nums
     * @return
     */
    public List<List<Integer>> threeSum(int[] nums) {
        if (null == nums || nums.length < 3) {
            return new ArrayList<>();
        }

        //首先进行排序，后面要进行二分查找
        Arrays.sort(nums);
        int len = nums.length;
        List<List<Integer>> res = new ArrayList<>();
        Set<String> set = new HashSet<>();
        for (int i = 0; i < len - 2; ++i) {
            for (int j = i + 1; j < len - 1; ++j) {
                int target = 0 - (nums[i] + nums[j]);
                //在剩下的位置二分法查找target
                int k = Arrays.binarySearch(nums, j + 1, len, target);
                if (k > 0) {
                    List<Integer> li = Arrays.asList(nums[i], nums[j], nums[k]);
                    Collections.sort(li);
                    String key = getKey(li);
                    if (!set.contains(key)) {
                        set.add(key);
                        res.add(li);
                    }
                }
            }
        }
        return res;
    }

    private String getKey(List<Integer> key) {
        return String.valueOf(key.get(0)) + key.get(1) + key.get(2);
    }

    /**
     * 解题思路：第二种解题思路，速度会更快，看代码逻辑就能够直接理解，需要注意的是，这种解法完全不需要去重处理，因为从代码逻辑
     * 上保证了不可能出现重复的结果
     *
     * @param nums
     * @return
     */
    public List<List<Integer>> threeSum2(int[] nums) {
        if (null == nums || nums.length < 3) {
            return new ArrayList<>();
        }

        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        int len = nums.length;
        for (int i = 0; i < len - 2; ++i) {
            //nums[i]没有发生变化，那么就意味着没有必要在继续以当前值为起点
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int target = -nums[i];
            int left = i + 1;
            int right = len - 1;

            //因为排好序了，所以可以从两端开始加速查找
            while (left < right) {
                //因为排好序了，发现最小的值比target还大，那么肯定就找不到目标的target
                if (nums[left] > target) {
                    break;
                }
                int tmp = nums[left] + nums[right];
                if (tmp == target) {
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    ++left;
                    //之所以可以这么做是因为left确定了，right也就确定了，所以相同的值没有意义
                    while (left < right && nums[left] == nums[left - 1]) {
                        ++left;
                    }

                    --right;
                    while (left < right && nums[right] == nums[right + 1]) {
                        --right;
                    }
                } else if (tmp > target) {
                    //此时需要缩小right来满足要求
                    --right;
                    while (left < right && nums[right] == nums[right + 1]) {
                        --right;
                    }
                } else {
                    //此时需要增大left来满足要求
                    ++left;
                    while (left < right && nums[left] == nums[left - 1]) {
                        ++left;
                    }
                }
            }
        }
        return res;
    }
}

File: go/double_pointer/15_3Sum.go
package main

import (
	"fmt"
	"slices"
)

/*
*
解题思路：
解法1： 使用三个指针遍历结果，最后在做一下去重即可，该解法给干超时了，用一个map记录每个数字出现的次数，循环的时候只循环i和j，到第三个数的时候从map中找，这样子可以稍微加速这个暴力解法，
但是还不是最优的
解法2：首先对数组进行排序，再选定一个数字，当前的位置记为i，然后j=i+1, k=len(nums)-1，如果nums[i]大于0，那么说明当前三个数的和不可能等于0，流程结束
如果当前的sum为0，那么记录结果，然后尝试对j++, k--，如果nums[j]==nums[j+1]说明当前的值存在重复，直接跳过该数字，k也是同理，以此类推，该解法可大大提升速度,
需要注意的是，当sum比0大的时候，这个时候需要移动j，如果比0小，需要移动k
*/
func main() {
	var (
		nums = []int{-1, 0, 1, 2, -1, -4}
	)
	result := threeSum(nums)
	result2 := threeSum2(nums)
	fmt.Println(result, "---------", result2) // [[-1,-1,2],[-1,0,1]]

	nums = []int{0, 1, 1}
	result = threeSum(nums)
	result2 = threeSum2(nums)
	fmt.Println(result, "---------", result2) // []

	nums = []int{0, 0, 0}
	result = threeSum(nums)
	result2 = threeSum2(nums)
	fmt.Println(result, "---------", result2) // [0,0,0]
}

func threeSum(nums []int) [][]int {
	if len(nums) < 3 {
		return nil
	}

	var (
		result    = make(map[[3]int]interface{}) // 使用map进行去除
		num2Count = make(map[int]int)
		i         = 0
		j         = 1
	)
	for _, val := range nums {
		num2Count[val]++
	}
	for ; i < len(nums)-2; i++ {
		for j = i + 1; j < len(nums)-1; j++ {
			target := 0 - (nums[i] + nums[j])
			count, ok := num2Count[target]
			if !ok {
				continue
			}
			needCount := 1
			if target == nums[i] {
				needCount++
			}
			if target == nums[j] {
				needCount++
			}
			if count >= needCount {
				key := [3]int{nums[i], nums[j], target}
				slices.Sort(key[:]) // 将数组转换为slice，可以进行排序，此时实际上slice引用的还是原来的数组
				result[key] = nil
			}
		}
	}
	res := make([][]int, 0)
	for key := range result {
		res = append(res, key[:])
	}
	return res
}

// threeSum2 更巧妙的一个解法，使用三个指针在有序的数组中快速找到答案
func threeSum2(nums []int) [][]int {
	if len(nums) < 3 {
		return nil
	}

	slices.Sort(nums)
	result := make([][]int, 0)
	// 保证最后能留下三个数
	for i := 0; i < len(nums)-2; i++ {
		if nums[i] > 0 {
			return result
		}
		// 注意，i也需要跳过重复的值，不然最后的结果中可能会出现重复值
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		j := i + 1
		k := len(nums) - 1
		for j < k && j < len(nums)-1 {
			sum := nums[i] + nums[j] + nums[k]
			if sum == 0 {
				result = append(result, []int{nums[i], nums[j], nums[k]})
				j++
				// 此时已经满足条件了，这里为了去除掉重复的数字，还要继续移动j和k
				for j < len(nums)-1 && nums[j] == nums[j-1] && j < k {
					j++
				}
				// 去除掉右边界重复的值
				k--
				for j < k && nums[k] == nums[k+1] {
					k--
				}
			} else if sum < 0 { // 结果比0小，需要增大sum
				j++
			} else { // 结果比0大，需要减小sum
				k--
			}
		}
	}
	return result
}
